### PR TITLE
adding optional body for HttpClient.delete request

### PR DESCRIPTION
**Pull request by gopal-jayaraman**
adding optional body for HTTP delete request. This new param added as an optional so won't break the existing code also provide the capability the send the body when and where it required.

## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

- [x] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:

## What is the current behavior?
The HTTP DELETE request does not support the request body.

Issue Number: #19438

## What is the new behavior?
adding optional body for HTTP delete request. This new param added as an optional so won't break the existing code also provide the capability the send the body when and where it required.

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

### CODE REVIEW

**Review by petebacondarwin — CHANGES_REQUESTED** [review]
This is looking good @gopal-jayaraman  - thanks. I'll give it a full review tomorrow.
In the meantime, could you squash the two commits into one and change the commit message to follow our [required syntax](https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit)?

Oh and you need to update the API golden file, which you can do by running:

```
yarn bazel run //packages/common:common_api.accept
```

File: packages/common/http/src/request.ts [hunk 1]
```diff
@@ -143,15 +142,15 @@ export class HttpRequest<T> {
    */
   readonly urlWithParams: string;
 
-  constructor(method: 'DELETE'|'GET'|'HEAD'|'JSONP'|'OPTIONS', url: string, init?: {
+  constructor(method: 'GET'|'HEAD'|'JSONP'|'OPTIONS', url: string, init?: {
```
Comment (petebacondarwin) [inline packages/common/http/src/request.ts:145]: This would be a breaking change.

**Review by petebacondarwin — COMMENTED** [review]
It is worth noting that a body (payload) is allowed on such requests as described in https://tools.ietf.org/html/rfc7231#section-4.3.5

>    A payload within a DELETE request message has no defined semantics; sending a payload body on a DELETE request might cause some existing implementations to reject the request.

I note that there is the same declaration for all the request types that do not normally expect a body.

---

Regarding this PR, the chosen signature, in `HttpClient`, for `delete()` is different both to the requests that always expected a body (e.g. `put()`, `post()`, etc) but also different to the generic `request()` method.

Given that a body for `DELETE` is an uncommon requirement, I would prefer that we just add it as an optional property on the `options` parameter, inline with how `request()` works. This would also make it simpler to pass through to the `request()` method without the need to call `addBody()`, and keep the signature as close to what it was before.

Furthermore, since providing a body does not appear to be specific to `DELETE`, I would also propose that we just allow `body` to be added as an optional property of `options` on all the request types that do not explicitly block it in RFC 7231. @alxhub ??

File: packages/common/http/src/request.ts (hunk 1 above)
Comment (gopal-jayaraman) [reply packages/common/http/src/request.ts]: fixed it here https://github.com/angular/angular/pull/41723, Please let me know if any concerns.

**Comment by AndrewKushnir** [comment]
Closing this PR in favor of #41723 (a copy of this one).

**Pull request closed without merging**